elements/lisk-tree/src/merkle_tree.ts

```diff
@@ -148,206 +144,73 @@ export class MerkleTree {
 				(leftNodeInfo as NodeInfo).layerIndex + 1,
 				(leftNodeInfo as NodeInfo).nodeIndex + 1,
 			);
-			appendPath.push(this.getNode(newBranchNode.hash));
+			appendPath.push(this.getNode(newBranchNode.hash) as NodeInfo);
 		}
 		this._root = appendPath[0].hash;
 		return this.root;
 	}
 
-	public generatePath(_queryData: ReadonlyArray<Buffer>): Path {
+	public generateProof(queryData: ReadonlyArray<Buffer>): Proof | undefined {
 		if (this._width === 1) {
-			return [];
+			return undefined;
 		}
-		const treeStructure = this._getStructure();
+		const treeStructure = this._getPopulatedStructure();
```

Comment:
this treeStucture is used in only one place and I think the same functionality can be achieved by just looking up `_hashToValueMap`.
Prefer to remove `_getPopulatedStructure` since it loops O(n) where n is 2 ** (height + 1) + 1

Reply:
`getPairLocation` only returns the coordinates of the pair node. This is done intentionally to keep the function a pure function since it is also used in `verifyProof`. Since we only have the coordinates of the pair node, it is necessary to maintain the `treeStructure` to get the pair node hash.